agent/src/protocols/udev/discovery_impl.rs

```diff
@@ -382,6 +436,126 @@ fn filter_by_remaining_udev_filters(
     mutable_devices
 }
 
+/// Recursively look up a device's hierarchy to see if it has an ancestor with a specified subsystem.
+pub fn device_has_parent_with_subsystem(device: &impl DeviceExt, subsystem: &str) -> bool {
+    match get_parent_with_subsystem(device, subsystem).unwrap() {
+        Some(_) => {
+            println!(
+                "true for subsystem {} and parent {:?}",
+                subsystem,
+                get_sysname(device)
+            );
+            true
+        }
+        None => {
+            println!(
+                "None for subsystem {} and parent {:?}",
+                subsystem,
+                get_sysname(device)
+            );
+            if let Some(parent) = get_parent(device) {
+                device_has_parent_with_subsystem(&parent, subsystem)
+            } else {
+                false
+            }
+        }
+    }
+}
+
+/// Recursively look up a device's hierarchy to see if it has an ancestor with a specified attribute.
+pub fn device_has_parent_with_attribute(
```

Comment:
Curious why the structure of this function is different from device_has_parent_with_subsystem?

Reply:
`device_has_parent_with_subsystem` is different from the other `S` fields because the rust udev library has support for a `parent_with_subsystem` function, while the others don't